File: code/stable/glassbr/src/python/ReadTable.py
## \file ReadTable.py
# \author Nikitha Krithnan and W. Spencer Smith
# \brief Provides a function for reading glass ASTM data
from __future__ import print_function
import sys
import math

## \brief Reads glass ASTM data from a file with the given file name
# \param filename name of the input file
# \param z_vector list of z values
# \param x_matrix lists of x values at different z values
# \param y_matrix lists of y values at different z values
def func_read_table(filename, z_vector, x_matrix, y_matrix):
    outfile = open("log.txt", "a")
    print("function func_read_table called with inputs: {", file=outfile)
    print("  filename = ", end='', file=outfile)
    print(filename, end='', file=outfile)
    print(", ", file=outfile)
    print("  z_vector = ", end='', file=outfile)
    print(z_vector, end='', file=outfile)
    print(", ", file=outfile)
    print("  x_matrix = ", end='', file=outfile)
    print(x_matrix, end='', file=outfile)
    print(", ", file=outfile)
    print("  y_matrix = ", end='', file=outfile)
    print(y_matrix, file=outfile)
    print("  }", file=outfile)
    outfile.close()
    
    linetokens = []
    lines = []
    infile = open(filename, "r")
    line = infile.readline().rstrip()
    linetokens = line.split(",")
    for stringlist_i in range(0, len(linetokens) // 1, 1):
        z_vector.append(float(linetokens[stringlist_i * 1 + 0]))
    outfile = open("log.txt", "a")
    print("var 'z_vector' assigned to ", end='', file=outfile)
    print(z_vector, end='', file=outfile)
    print(" in module ReadTable", file=outfile)
    outfile.close()
    lines = infile.readlines()
    for i in range(0, len(lines), 1):
        linetokens = lines[i].split(",")
        x_matrix_temp = []
        y_matrix_temp = []
        for stringlist_i in range(0, len(linetokens) // 2, 1):
            x_matrix_temp.append(float(linetokens[stringlist_i * 2 + 0]))
            y_matrix_temp.append(float(linetokens[stringlist_i * 2 + 1]))
        x_matrix.append(x_matrix_temp)
        y_matrix.append(y_matrix_temp)
    outfile = open("log.txt", "a")
    print("var 'x_matrix' assigned to ", end='', file=outfile)
    print(x_matrix, end='', file=outfile)
    print(" in module ReadTable", file=outfile)
    outfile.close()
    outfile = open("log.txt", "a")
    print("var 'y_matrix' assigned to ", end='', file=outfile)
    print(y_matrix, end='', file=outfile)
    print(" in module ReadTable", file=outfile)
    outfile.close()
    infile.close()

File: code/stable/GlassBR/src/swift/ReadTable.swift
/** ReadTable.swift
    Provides a function for reading glass ASTM data
    - Authors: Nikitha Krithnan and W. Spencer Smith
*/
import Foundation

/** Reads glass ASTM data from a file with the given file name
    - Parameter filename: name of the input file
    - Parameter z_vector: list of z values
    - Parameter x_matrix: lists of x values at different z values
    - Parameter y_matrix: lists of y values at different z values
*/
func read_table(_ filename: String, _ z_vector: inout [Double], _ x_matrix: inout [[Double]], _ y_matrix: inout [[Double]]) throws -> Void {
    var outfile: FileHandle
    do {
        outfile = try FileHandle(forWritingTo: FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!.appendingPathComponent("log.txt"))
        try outfile.seekToEnd()
    } catch {
        throw "Error opening file."
    }
    do {
        try outfile.write(contentsOf: Data("function read_table called with inputs: {".utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data("  filename = ".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data(filename.utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data(", ".utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data("  z_vector = ".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data(z_vector.description.utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data(", ".utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data("  x_matrix = ".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data(x_matrix.description.utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data(", ".utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data("  y_matrix = ".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data(y_matrix.description.utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data("  }".utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.close()
    } catch {
        throw "Error closing file."
    }
    
    var infile: URL
    var line: String
    var linetokens: [String] = []
    var lines: [String] = []
    infile = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!.appendingPathComponent(filename)
    var goolContents: [[String]]
    do {
        goolContents = try String(contentsOf: infile).components(separatedBy: "\n").map({(l: String) -> [String] in l.components(separatedBy: " ")})
    } catch {
        throw "Error reading from file."
    }
    var goolLine: [String]
    goolLine = [Int](stride(from: 0, to: goolContents[0].count, by: 1)).map({(i: Int) -> String in goolContents[0][i]})
    line = goolLine.joined(separator: " ")
    linetokens = line.components(separatedBy: ",")
    for stringlist_i in [Int](stride(from: 0, to: linetokens.count / 1, by: 1)) {
        z_vector.append(Double(linetokens[stringlist_i * 1 + 0])!)
    }
    do {
        outfile = try FileHandle(forWritingTo: FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!.appendingPathComponent("log.txt"))
        try outfile.seekToEnd()
    } catch {
        throw "Error opening file."
    }
    do {
        try outfile.write(contentsOf: Data("var 'z_vector' assigned ".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data(z_vector.description.utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data(" in module ReadTable".utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.close()
    } catch {
        throw "Error closing file."
    }
    goolContents = [Int](stride(from: 2, to: goolContents.count, by: 1)).map({(i: Int) -> [String] in goolContents[i]})
    lines = goolContents.map({(l: [String]) -> String in l.joined(separator: " ")})
    for i in [Int](stride(from: 0, to: lines.count, by: 1)) {
        linetokens = lines[i].components(separatedBy: ",")
        var x_matrix_temp: [Double] = []
        var y_matrix_temp: [Double] = []
        for stringlist_i in [Int](stride(from: 0, to: linetokens.count / 2, by: 1)) {
            x_matrix_temp.append(Double(linetokens[stringlist_i * 2 + 0])!)
            y_matrix_temp.append(Double(linetokens[stringlist_i * 2 + 1])!)
        }
        x_matrix.append(x_matrix_temp)
        y_matrix.append(y_matrix_temp)
    }
    do {
        outfile = try FileHandle(forWritingTo: FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!.appendingPathComponent("log.txt"))
        try outfile.seekToEnd()
    } catch {
        throw "Error opening file."
    }
    do {
        try outfile.write(contentsOf: Data("var 'x_matrix' assigned ".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data(x_matrix.description.utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data(" in module ReadTable".utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.close()
    } catch {
        throw "Error closing file."
    }
    do {
        outfile = try FileHandle(forWritingTo: FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!.appendingPathComponent("log.txt"))
        try outfile.seekToEnd()
    } catch {
        throw "Error opening file."
    }
    do {
        try outfile.write(contentsOf: Data("var 'y_matrix' assigned ".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data(y_matrix.description.utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data(" in module ReadTable".utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.close()
    } catch {
        throw "Error closing file."
    }
}
